File: android/src/main/java/com/webee/react/RNXRPC.java

package com.webee.react;

import android.os.Bundle;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.module.annotations.ReactModule;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import rx.Observable;
import rx.subjects.AsyncSubject;
import rx.subjects.PublishSubject;

import static com.webee.react.RNXRPCClient.requests;

@ReactModule(name = "XRPC")
public class RNXRPC extends ReactContextBaseJavaModule {
    public static final String XRPC_EVENT_CALL = "__XRPC.E.CALL";
    public static final String XRPC_EVENT_REPLY = "__XRPC.E.REPLY";
    public static final String XRPC_EVENT_REPLY_ERROR = "__XRPC.E.REPLY_ERROR";
    public static final String XRPC_EVENT_EVENT = "__XRPC.E.EVENT";
    private static final Map<String, Object> constants;
    private static final Map<String, Procedure> procedures = new ConcurrentHashMap<>();
    private static final PublishSubject<Event> eventSubject = PublishSubject.create();

    static {
        constants = new HashMap<>();
        constants.put("EVENT_CALL", XRPC_EVENT_CALL);
        constants.put("EVENT_REPLY", XRPC_EVENT_REPLY);
        constants.put("EVENT_REPLY_ERROR", XRPC_EVENT_REPLY_ERROR);
        constants.put("EVENT_EVENT", XRPC_EVENT_EVENT);
    }

    public RNXRPC(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @Override
    public String getName() {
        return "XRPC";
    }

    @Override
    public Map<String, Object> getConstants() {
        return constants;
    }

    @ReactMethod
    public void emit(final String event, final ReadableArray args) {
        switch (event) {
            case XRPC_EVENT_REPLY:
                handleCallReply(args);
                break;
            case XRPC_EVENT_REPLY_ERROR:
                handleCallReplyError(args);
                break;
            case XRPC_EVENT_CALL:
                handleCall(args);
                break;
            case XRPC_EVENT_EVENT:
                handleEvent(args);
            default:
                break;
        }
    }

    private void handleEvent(ReadableArray xargs) {
        String event = xargs.getString(0);
        ReadableArray args = xargs.getArray(1);
        ReadableMap kwargs = xargs.getMap(2);
        eventSubject.onNext(new Event(event, args, kwargs));
    }

    private void handleCall(ReadableArray xargs) {
        String rid = xargs.getString(0);
        String proc = xargs.getString(1);
        ReadableArray args = xargs.getArray(2);
        ReadableMap kwargs = xargs.getMap(3);
        Procedure f = procedures.remove(proc);
        if (f == null) {
            return;
        }
        // TODO:
    }

    private void handleCallReply(ReadableArray xargs) {
        String rid = xargs.getString(0);
        AsyncSubject<Reply> replySubject = RNXRPCClient.requests.remove(rid);
        if (replySubject == null) {
            return;
        }

        ReadableArray args = xargs.getArray(1);
        ReadableMap kwargs = xargs.getMap(2);

        replySubject.onNext(new Reply(args, kwargs));
        replySubject.onCompleted();
    }

    private void handleCallReplyError(ReadableArray xargs) {
        String rid = xargs.getString(0);
        AsyncSubject<Reply> replySubject = requests.remove(rid);
        if (replySubject == null) {
            return;
        }

        String error = xargs.getString(1);

        int s = xargs.size();
        ReadableArray args = null;
        ReadableMap kwargs = null;
        if (s > 2) {
            args = xargs.getArray(2);
        }
        if (s > 3) {
            kwargs = xargs.getMap(3);
        }

        replySubject.onError(new XRPCError(error, args, kwargs));
    }

    public static Observable<Event> event() {
        return eventSubject.asObservable();
    }

    public static void register() {
        // TODO: register procedure
    }

    public interface Procedure {
        void run(ReadableArray args, ReadableMap kwargs, ReplyAPI reply);
    }

    public interface ReplyAPI {
        void onReply(final Object[] args, final Bundle kwargs);
        void onError(String err, final Object[] args, final Bundle kwargs);
    }
}

File: ios/ReactNativeXRPC/RNXRPC.m
//
//  XRPC.m
//  RNXRPC
//
//  Created by webee on 16/10/20.
//  Copyright © 2016年 Facebook. All rights reserved.
//

#import "RNXRPC.h"
#import <ReactiveObjC/RACSubject.h>
#import <React/RCTLog.h>

NSString* const XRPC_EVENT = @"__XRPC__";
NSInteger const XRPC_EVENT_CALL = 0;
NSInteger const XRPC_EVENT_REPLY = 1;
NSInteger const XRPC_EVENT_REPLY_ERROR = 2;
NSInteger const XRPC_EVENT_EVENT = 3;

@implementation RNXRPC {
    NSDictionary* _extraConstants;
}
static RACSubject *__eventSubject;
static NSMutableDictionary<NSString*, RNXRPCOnReplyBlock>* requests;
static NSLock* reqLock;

@synthesize bridge = _bridge;

+ (void)initialize {
    __eventSubject = [RACSubject subject];

    requests = [NSMutableDictionary new];
    reqLock = [[NSLock alloc] init];
}

- (id) init {
    return [self initWithExtraConstants:nil];
}

- (id) initWithExtraConstants:(NSDictionary*)constants {
    if (self = [super init]) {
        if (constants != nil) {
            _extraConstants = constants;
        } else {
            _extraConstants = @{};
        }
    }
    return self;
}

RCT_EXPORT_MODULE(XRPC);

- (dispatch_queue_t)methodQueue {
    return dispatch_queue_create("com.webee.react.XRPCStorageQueue", DISPATCH_QUEUE_SERIAL);
}

- (NSDictionary *)constantsToExport {
    return @{ @"_XRPC_EVENT": XRPC_EVENT,
              @"_EVENT_CALL": @(XRPC_EVENT_CALL),
              @"_EVENT_REPLY": @(XRPC_EVENT_REPLY),
              @"_EVENT_REPLY_ERROR": @(XRPC_EVENT_REPLY_ERROR),
              @"_EVENT_EVENT": @(XRPC_EVENT_EVENT),
              @"C": _extraConstants
              };
}

RCT_EXPORT_METHOD(emit:(NSInteger)event args:(NSArray *)args) {
    switch (event) {
        case XRPC_EVENT_REPLY:
            [self handleCallReply:args];
            break;
        case XRPC_EVENT_REPLY_ERROR:
            [self handleCallReplyError:args];
            break;
        case XRPC_EVENT_EVENT:
            [self handleEvent:args];
            break;
        default:
            break;
    }
}

RCT_EXPORT_METHOD(call:(NSString *)proc args:(NSArray *)args kwargs:(NSDictionary *)kwargs
            resolver:(RCTPromiseResolveBlock)resolve
            rejecter:(RCTPromiseRejectBlock)reject) {
    // TODO: 接受call
    RCTLog(@"call %@, %@, %@", proc, args, kwargs);
}

- (void) handleEvent:(NSArray*)xargs {
    NSString* event = xargs[0];
    NSArray* args = xargs[1];
    NSDictionary* kwargs = xargs[2];
    [__eventSubject sendNext:[[RNXRPCEvent alloc] initWithArgs:_bridge event:event args:args kwargs:kwargs]];
}

- (void) handleCallReply:(NSArray*)xargs {
    NSString* rid = xargs[0];
    [reqLock lock];
    RNXRPCOnReplyBlock onReply = requests[rid];
    if (onReply == nil) {
        return;
    }
    [requests removeObjectForKey:rid];
    [reqLock unlock];
    NSArray* args = xargs[1];
    NSDictionary* kwargs = xargs[2];
    
    onReply([[RNXRPCReply alloc] initWithArgs:args kwargs:kwargs], nil);
}

- (void) handleCallReplyError:(NSArray*)xargs {
    NSString* rid = xargs[0];
    [reqLock lock];
    RNXRPCOnReplyBlock onReply = requests[rid];
    if (onReply == nil) {
        return;
    }
    [requests removeObjectForKey:rid];
    [reqLock unlock];
    NSString* err = xargs[1];
    NSArray* args = xargs[2];
    NSDictionary* kwargs = xargs[3];
    onReply(nil, [[RNXRPCError alloc] initWithArgs:err args:args kwargs:kwargs]);
}

+ (nonnull RACSignal<RNXRPCEvent *> *)event {
    return (RACSignal<RNXRPCEvent *> *) __eventSubject;
}

+ (void) request:(NSString*)rid onReply:(RNXRPCOnReplyBlock)onReply {
    [reqLock lock];
    requests[rid] = onReply;
    [reqLock unlock];
}
@end
